File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
      int left = 0; 
      int r = heights.size()-1; 
      int max = 0;
      while (left < r) {
        int width = r - left; 
        int height = min(heights[left], heights[r]); 
        int area = width * height;
        if (area > max) {
            max = area; 
        }
        // int maxh = max(height[left], height[r]);
        if (heights[left] > heights[r]) {
            r--; 
        }
        else if (heights[left] < heights[r]) {
            left++; 
        }  
        else {
            if (heights[left + 1] > heights[r - 1]) {
                left++; 
            }
            else {
                r--; 
            }
        }
      } 
      return max; 
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # 0,2,0,3,1,0,1,3,2,1
        # 9

        # 2 pointers 
        # two bars where l < r
        # for all of these indexes in between < height of the smallest bar 
        # area = for each index in between the two bars, height of the smallest bar - height of the bar at the curr index

        max_area = 0

        l = 0
        r = len(height) - 1

        maxleft = height[l]
        maxRight = height[r]

        while (l < r):
            if maxleft < maxRight:
                l += 1
                maxleft = max(maxleft, height[l])
                max_area += maxleft - height[l]
            else:
                r -= 1
                maxRight = max(maxRight, height[r])
                max_area += maxRight - height[r]
               


        return max_area


        